Rework Fluent file path and yaml references

pchote opened the pull request:
Followup to #21602 and implements most of the remaining points from the previous Discord discussion.

The motivation for these changes is to remove the magic file path replacements and the specific references to "en" in the file paths. The exact shape of the future ingame translation system is still up for debate, but i'm fairly confident that it will end up as some variation of one of the following approaches (likely the first):

1. Dynamic file overrides for mod-defined paths - e.g. via a downloadable translation bundle that contains a manifest saying "remap file `ra|fluent/rules.ftl` to `langbundle|rules.ftl`. In this case it would be incorrect to specify `en` in the mod-defined paths because the files that they load may not be in english.

2. Explicitly defined file paths for each language at the point of usage - in which case we will need to update all of these file references when future languages are supported, meaning that nothing is lost by removing "en" from the paths in the meantime.

The incomplete user-level language selection is replaced by a mod-level language key. This enables modders and downstream repackagers to ship non-english languages and not have Fluent misbehave due to assuming the "en" culture.

For the time being, this PR updates the logic added in #21601 for passing strings to the content installer, but I intend to file a third PR in this series that will remove a lot of that magic in favour of an explicit mod.yaml definition for each content installer mod.

PunkPun reviewed (commented):
`UserInterface.String()` is definitely weird, we should do something different. At the minimum `UserInterface.GetString()`

I recon the best might be `FluentProvider.GetString()`?

PunkPun commented on OpenRA.Game/Manifest.cs:
```diff
@@ -83,17 +83,19 @@ public readonly string[]
 		public readonly string[] VideoFormats = Array.Empty<string>();
-		public readonly bool AllowUnusedTranslationsInExternalPackages = true;
+
+		public readonly string FluentStringsLanguage = "en";
```
this doesn't seem right. It is used just to create CultureInfo

pchote replied:
We could rename to `FluentCulture`?

PunkPun replied:
I feel like it shouldn't exist in the manifest. If the language itself isn't defined in the manifest, so the culture shouldn't either

pchote replied:
How should this be specified if somebody wants to e.g. repackage RA in german, or OpenHV in chinese?

PunkPun replied:
Realistically fonts shouldn't be defined here as well, perhaps only the font name the rest of the engine uses. For short term we could have `FluentCulture` and in the manifest add a `TODO`

pchote replied:
Fonts are going to be tricky - everything that draws text is going to need to work out whether the message is in the user-selected language or using the fallback english message and update its font as needed.

PunkPun reviewed (commented):
while `FluentStrings` is a passable name, I feel like strings is the wrong term here. We are expecting .ftl files, not strings

FTL stands for Fluent Translation List, and it's the expected format of the field. Perhaps a better name would be `FluentTranslationFiles`, following the already existing convention.

pchote commented:
Actually, the various nodes in map.yaml do expect raw file contents as child nodes, with the ability to reference external files being an odd implementation detail.

The fluent docs define that the basic unit of translation as a "message", so we should probably be replacing all uses of GetString etc with GetMessage.

pchote commented:
> while FluentStrings is a passable name, I feel like strings is the wrong term here. We are expecting .ftl files, not strings

The terminology in Linguini and our wrapping of that is to use "Bundle" to refer to a collection of messages. So would `FluentBundles` be an acceptable replacement?

pchote commented:
I'll wait for #21623 and #21570 before rebasing.

pchote committed:
Rename Fluent *GetString methods to GetMessage.

pchote committed:
Use nameof() when reflecting over Map field names.

pchote committed:
Reorganize FluentBundle ctors to allow inline text.

pchote committed:
Add map support for inline base64 fluent messages.

This enables the RC to parse and share custom messages
as part of the map's custom rules without any additional
API changes.

pchote committed:
Simplify method name in CheckFluentReferences.

pchote committed:
Rename Translations yaml keys to FluentMessages.

pchote committed:
Rename Lua UserInterface.Translate to .GetFluentMessage.

pchote committed:
Remove "en" from fluent file paths.

pchote commented:
Rebased on #21623 and reworked based on feedback above and in Discord.

This now also fixes #21618 by updating the `--map-rules` utility command that is used by the RC to populate the "rules" yaml block to include inline ftl. e.g. for allies-02, it now includes
```
FluentMessages: ra|fluent/lua.ftl, ra|fluent/campaign.ftl
	base64: ZHJvcGRvd24tZGlmZmljdWx0eSA9CiAgICAubGFiZWwgPSBEaWZmaWN1bHR5CiAgICAuZGVzY3JpcHRpb24gPSBUaGUgZGlmZmljdWx0eSBvZiB0aGUgbWlzc2lvbgoKb3B0aW9ucy1kaWZmaWN1bHR5ID0KICAgIC5lYXN5ID0gRWFzeQogICAgLm5vcm1hbCA9IE5vcm1hbAogICAgLmhhcmQgPSBIYXJkCiAgICAudG91Z2ggPSBSZWFsIHRvdWdoIGd1eQo=
```
and `MapPreview`/`Map` will decode this to include in the map bundle.

Mailaender commented on OpenRA.Mods.Common/Scripting/Global/UserInterfaceGlobal.cs:
```diff
@@ -35,7 +35,7 @@ public void SetMissionText(string text, Color? color = null)
 		[Desc("Formats a language string for a given string key defined in the language files (*.ftl). " +
 			"Args can be passed to be substituted into the resulting message.")]
-		public string Translate(string key, [ScriptEmmyTypeOverride("{ string: any }")] LuaTable args = null)
+		public string FluentMessage(string key, [ScriptEmmyTypeOverride("{ string: any }")] LuaTable args = null)
```
I think the previous API name is explaining things better to scripters.

PunkPun replied:
I think GetFluentMessage is good enough

pchote replied:
Ok, renamed.

Mailaender commented on OpenRA.Game/Map/MapPreview.cs:
```diff
@@ -335,7 +355,7 @@ public MapPreview(Map map, ModData modData)
 			{
 				{ "Rules", map.RuleDefinitions },
-				{ "Translations", map.TranslationDefinitions },
+				{ "FluentMessages", map.FluentMessageDefinitions },
```
This is also worse. We don't call it `MiniYamlRules` but just `Rules`. Please keep this as is for consistency.

PunkPun replied:
It is definitely confusing. I almost feel like just having it be `Fluent` would be better

PunkPun replied:
https://github.com/OpenRA/OpenRA/pull/21622#pullrequestreview-2375136239

pchote replied:
> In Fluent, the basic unit of translation is called a message. Messages are containers for information. You use messages to identify, store, and recall translation information to be used in the product.

This node is listing filenames containing/inline definitions of fluent messages. This is the same as "Sequences" listing filenames containing/inline definitions of sequences, etc.

There is no logic here to handle multiple languages, which I feel is a basic requirement for calling something "Translations".

Mailaender replied:
The definition already says that it is `TranslationMessages`. People who don't know that @projectfluent is a translation framework will be irritated. Ideally you should revert the renamings from #21602 as well.

pchote replied:
That would conflict with the naming of things in the actual translation system that manages the sets of language-specific fluent / sound / sprite files.

PunkPun replied:
On a yaml level I think this is the only place it is called fluent? So for non-power users both the concept of fluent and messages might be foreign

PunkPun replied:
for more advanced modders the other place will be https://github.com/OpenRA/OpenRA/pull/21622#discussion_r1807804555. I think we should make sure these are named well.

Mailaender commented on OpenRA.Game/Manifest.cs:
```diff
@@ -83,17 +83,20 @@ public readonly string[]
 		public readonly int FontSheetSize = 512;
 		public readonly int CursorSheetSize = 512;
 
+		// TODO: This should be controlled by a user-selected translation bundle!
+		public readonly string FluentCulture = "en";
```
No, it should be controlled by a user settings like it already is.

PunkPun replied:
it should be controlled by the bundles. The user should be able to select the language from installed languages eventually yeah. It should give the pop-up window to reload the game same way changing display size does

PunkPun commented on OpenRA.Mods.Common/Lint/CheckFluentReferences.cs:
```diff
@@ -80,29 +78,24 @@ void ILintPass.Run(Action<string> emitError, Action<string> emitWarning, ModData
+
+			foreach (var group in usedKeys.KeysWithContext)
 			{
-				Console.WriteLine($"Testing language: {language}");
```
We should keep a console print to be consistent with other major lint test

pchote replied:
Ok, done.

Mailaender commented on OpenRA.Game/FluentProvider.cs:
```diff
@@ -25,9 +27,25 @@ public static void Initialize(ModData modData, IReadOnlyFileSystem fileSystem)
+					{
+						var builder = new StringBuilder();
+						foreach (var node in map.FluentMessageDefinitions.Nodes)
+							if (node.Key == "base64")
+								builder.Append(Encoding.UTF8.GetString(Convert.FromBase64String(node.Value.Value)));
```
Why are you encoding text as `base64`?

pchote replied:
The fluent grammar overlaps with MiniYaml, so cannot be directly included as yaml nodes without a horrible level of escaping. This feature is implemented for the RC, humans aren't meant to be manually constructing these (unless they *really* want to).

PunkPun replied:
https://github.com/OpenRA/OpenRA/pull/21622#issuecomment-2423927317

Mailaender replied:
I don't think the resource center should be streaming translations. It should serve a ZIP that is then extracted and read when downloaded.

pchote replied:
It isn't streaming translations. It's including the fluent messages that are defined by the map, and are needed to correctly display the lobby options.

Otherwise, as a best case scenario, you end up with things like this:
<img width="225" alt="Screenshot 2024-10-20 at 13 56 47" src="https://github.com/user-attachments/assets/11ced1fb-9283-48c8-a4b7-ab14eb3a4da0">

pchote committed:
Remove magic ftl file naming.

PunkPun reviewed (approved):
This changes done here should be good enough for playtest

PunkPun commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/30556859ea82f0e17579346d53c986e723fb9179...ec7d5e1ea31ea84c67e5fdf421508c31f648ab7d)
